core/src/test/java/org/apache/calcite/util/SealableMapTest.java

```diff
@@ -0,0 +1,247 @@
+    });
+
+    ensureFailure(() -> {
+      map.values().clear();
+      return "values mutate";
+    });
+
+    ensureFailure(() -> {
+      map.entrySet().clear();
+      return "entrySet mutate";
+    });
+
+    ensureFailure(() -> {
+      map.replaceAll((k, v) -> v);
+      return "replaceAll";
+    });
+
+    ensureFailure(() -> {
+      map.putIfAbsent("foo", "bar");
+      return "putIfAbsent";
+    });
+
+    ensureFailure(() -> {
+      map.remove("foo", "bar");
+      return "remove(k,v)";
+    });
+
+    ensureFailure(() -> {
+      map.replace("foo", "bar", "baz");
+      return "replace(k,o_v,n_v)";
+    });
+
+    ensureFailure(() -> {
+      map.replace("foo", "baz");
+      return "replace(k,n_v)";
+    });
+
+    ensureFailure(() -> {
+      map.computeIfAbsent("foo", ignored -> "bar");
+      return "computeIfAbsent";
+    });
+
+    ensureFailure(() -> {
+      map.computeIfPresent("foo", (ignored1, ignored2) -> "bar");
+      return "computeIfPresent";
+    });
+
+    ensureFailure(() -> {
+      map.compute("foo", (ignored1, ignored2) -> "bar");
+      return "compute";
+    });
+
+    ensureFailure(() -> {
+      map.merge("foo", "bar", (ignored1, ignored2) -> "baz");
+      return "merge";
+    });
+  }
+
+  private void ensureUnsealed(SealableMap<String, String> map, Map<String, String> backingMap) {
+    assertFalse(map.isSealed());
+    ensureReadable(map, backingMap);
+
+    Map<String, String> backingMapCopy = new HashMap<>(backingMap);
+
+    String key = "@@key@@";
+    String value1 = "@@value1@@";
+    String value2 = "@@value2@@";
+
+    map.put(key, value1);
+    assertEquals(map.get(key), value1);
+    map.remove(key);
+    assertFalse(map.containsKey(key));
+
+    map.clear();
+    map.putAll(backingMapCopy);
+    assertEquals(map.entrySet(), backingMapCopy.entrySet());
+
+    map.keySet().clear();
+    map.putAll(backingMapCopy);
+
+    map.values().clear();
+    map.putAll(backingMapCopy);
+
+    map.entrySet().clear();
+    map.putAll(backingMapCopy);
+
+    map.replaceAll((k, v) -> value1);
+    map.values().forEach(value -> assertEquals(value, value1));
+    map.clear();
+    map.putAll(backingMapCopy);
+
+    map.putIfAbsent(key, value1);
+    assertEquals(map.get(key), value1);
+    map.replace(key, value1, value2);
+    assertEquals(map.get(key), value2);
+    map.replace(key, value1);
+    assertEquals(map.get(key), value1);
+    map.remove(key, value1);
+    assertFalse(map.containsKey(key));
+
+    map.computeIfAbsent(key, k -> value1);
+    assertEquals(map.get(key), value1);
+    map.computeIfPresent(key, (k, v) -> value2);
+    assertEquals(map.get(key), value2);
+    map.compute(key, (k, v) -> value1);
+    assertEquals(map.get(key), value1);
+    map.merge(key, value1, (k, v) -> null);
+    assertFalse(map.containsKey(key));
+
+    // clean up
+    map.clear();
+    map.putAll(backingMapCopy);
+  }
+
+  private <K, V> void ensureReadable(Map<K, V> map, Map<K, V> backingMap) {
```

Comment:
this is not checking if the map is readable right? this method is performing map equality; if so, lets change the method name to reflect that?

Reply:
Its checking all the "read" methods -- the equality checks are to make sure the read methods are actually returning data from the delegates. So this is #1 making sure map is readable #2 data returned is actually from delegate. I'll make this `ensureReadableAndDataMatchesDelegate`

Reply:
makes sense